lib/bandit/http1/adapter.ex

```diff
@@ -288,8 +247,31 @@ defmodule Bandit.HTTP1.Adapter do
 
   def read_req_body(%__MODULE__{}, _opts), do: raise(Bandit.BodyAlreadyReadError)
 
-  @dialyzer {:no_improper_lists, do_read_chunk: 5}
-  defp do_read_chunk(socket, buffer, body, read_size, read_timeout) do
+  @dialyzer {:no_improper_lists, do_read_content_length_body: 4}
+  defp do_read_content_length_body(socket, buffer, body_remaining, opts) do
+    max_desired_bytes = Keyword.get(opts, :length, 8_000_000)
+
+    if byte_size(buffer) >= max_desired_bytes || body_remaining == 0 do
+      # We can satisfy the read request entirely from our buffer
+      bytes_to_return = min(max_desired_bytes, byte_size(buffer))
+      <<to_return::binary-size(bytes_to_return), rest::binary>> = buffer
+      {:ok, to_return, rest, body_remaining}
+    else
+      # We need to read off the wire
+      bytes_to_read = min(max_desired_bytes - byte_size(buffer), body_remaining)
+      read_size = Keyword.get(opts, :read_length, 1_000_000)
+      read_timeout = Keyword.get(opts, :read_timeout)
+
+      with {:ok, iolist} <- read(socket, bytes_to_read, [], read_size, read_timeout) do
+        to_return = IO.iodata_to_binary([buffer | iolist])
+        body_remaining = body_remaining - (byte_size(to_return) - byte_size(buffer))
```

Comment:
Could you run tests please with:
```
        unquote(
          if Mix.env() in [:dev, :test] do
            quote do
              if var!(body_remaining) < 0 do
                raise "body_remaining < 0: #{var!(body_remaining)}"
              end
            end
          end
        )
```

just to be sure that the counter never go below 0

Reply:
Yep - this is a loose end I'm aware of and was planning on tackle as part of #174. I'll add a failing test that exercises this & we can skip it for now (I'm planning on tackling 174 later this afternoon so it won't be long)

Reply:
Ok, no problem then. Let me know please if I can assist you.